src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectableFieldNode.cs

```diff
@@ -0,0 +1,105 @@
+        {
+            Debug.Assert(!field.OwningType.IsCanonicalSubtype(CanonicalFormKind.Any)
+                || field.OwningType.ConvertToCanonForm(CanonicalFormKind.Specific) == field.OwningType);
+            _field = field;
+        }
+
+        public FieldDesc Field => _field;
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory factory)
+        {
+            Debug.Assert(!factory.MetadataManager.IsReflectionBlocked(_field.GetTypicalFieldDefinition()));
+
+            DependencyList dependencies = new DependencyList();
+            factory.MetadataManager.GetDependenciesDueToReflectability(ref dependencies, factory, _field);
+
+            // No runtime artifacts needed if this is a generic definition or literal field
+            if (_field.OwningType.IsGenericDefinition || _field.IsLiteral)
+            {
+                return dependencies;
+            }
+
+            FieldDesc typicalField = _field.GetTypicalFieldDefinition();
+            if (typicalField != _field)
+            {
+                // Ensure we consistently apply reflectability to all fields sharing the same definition.
+                // Bases for different instantiations of the field have a conditional dependency on the definition node that
+                // brings a ReflectableField of the instantiated field if it's necessary for it to be reflectable.
+                dependencies.Add(factory.ReflectableField(typicalField), "Definition of the reflectable field");
+            }
+
+            // Runtime reflection stack needs to see the type handle of the owning type
+            dependencies.Add(factory.MaximallyConstructableType(_field.OwningType), "Instance base of a reflectable field");
+
+            // Root the static base of the type
+            if (_field.IsStatic && !_field.OwningType.IsCanonicalSubtype(CanonicalFormKind.Any))
+            {
+                // Infrastructure around static constructors is stashed in the NonGC static base
+                bool needsNonGcStaticBase = factory.PreinitializationManager.HasLazyStaticConstructor(Field.OwningType);
+
+                if (_field.HasRva)
+                {
+                    // No reflection access right now
+                }
+                else if (_field.IsThreadStatic)
+                {
+                    dependencies.Add(factory.TypeThreadStaticIndex((MetadataType)_field.OwningType), "Threadstatic base of a reflectable field");
+                }
+                else if (_field.HasGCStaticBase)
+                {
+                    dependencies.Add(factory.TypeGCStaticsSymbol((MetadataType)_field.OwningType), "GC static base of a reflectable field");
+                }
+                else
+                {
+                    dependencies.Add(factory.TypeNonGCStaticsSymbol((MetadataType)_field.OwningType), "NonGC static base of a reflectable field");
+                    needsNonGcStaticBase = false;
+                }
```

Comment:
Why not just remove this branch and let the if immediately below do this? It seems to be doing the same thing.

Reply:
It's part of a if/else cascade - we only need to add the `TypeNonGCStaticsSymbol` if there's a class constructor, or the field is non-GC static. There's no quick way to check if a field is non-GC static - one has to ask the "is it RVA/ThreadStatic/GCStatic" questions first.

The best we could do here is change this `else` block to `needsNonGcStaticBase = true;` and delete the `dependencies.Add` line but then the reason string will be wrong.